vminfo_parser/vmdata.py

```diff
@@ -67,54 +82,80 @@ def set_column_headings(self: t.Self) -> None:
         if best_match is None:
             raise ValueError("No matching header set found")
 
-        self.column_headers = const.COLUMN_HEADERS[best_match].copy()
-        missing_headers = [header for header in self.column_headers.values() if header not in self.df.columns]
-        self.column_headers["unitType"] = "GB" if best_match == "VERSION_1" else "MB"
-
         LOGGER.debug(f"Using VERSION_{best_match} as the closest match.")
 
+        self.column_headers = const.COLUMN_HEADERS[best_match].copy()
+        self.unit_type = "GiB" if best_match == "VERSION_1" else "MiB"
+
+        missing_headers = [header for header in self.column_headers.values() if header not in self.df.columns]
         if missing_headers:
             LOGGER.critical("The following headers are missing: %s", missing_headers)
             exit()
 
-    def add_extra_columns(self: t.Self) -> None:
+    def _set_os_columns(self: t.Self) -> None:
+        """
+        Add os name and version columns to dataframe.
+
+        Checks to see if output columns exist in dataframe and returns if so.
+
+        Returns:
+            None
+        """
+        if all(col in self.df.columns for col in const.EXTRA_COLUMNS_DEST):
+            LOGGER.info("All columns already exist")
+            return None
+
         primary_os_column = self.column_headers.get("operatingSystemFromVMTools")
         secondary_os_column = self.column_headers.get("operatingSystemFromVMConfig")
 
-        combined_os_column = "combined_operating_system"
-        self.df[combined_os_column] = self.df[secondary_os_column].where(
-            self.df[primary_os_column].isnull(), self.df[primary_os_column]
+        combined_os: pd.Series = self.df[primary_os_column].fillna(self.df[secondary_os_column])
+
+        # Set "OS Name", "OS Version", "Architecture" with regex match of combined_os
+        self.df[const.EXTRA_COLUMNS_DEST] = (
+            # Parse as None Windows OS
+            combined_os.str.extract(const.EXTRA_COLUMNS_NON_WINDOWS_REGEX)
+            # If no match, parse as Windows Server
+            .fillna(combined_os.str.extract(const.EXTRA_COLUMNS_WINDOWS_SERVER_REGEX))
+            # If no match, parse as Windows Desktop
+            .fillna(combined_os.str.extract(const.EXTRA_COLUMNS_WINDOWS_DESKTOP_REGEX, flags=re.IGNORECASE))
         )
 
-        if not all(col in self.df.columns for col in const.EXTRA_COLUMNS_DEST):
-            self.df[const.EXTRA_COLUMNS_DEST] = self.df[combined_os_column].str.extract(
-                const.EXTRA_COLUMNS_NON_WINDOWS_REGEX
-            )
-            self.df[const.EXTRA_WINDOWS_SERVER_COLUMNS] = self.df[combined_os_column].str.extract(
-                const.EXTRA_COLUMNS_WINDOWS_SERVER_REGEX
```

Comment:
The previous logic was could be described as 

1. check if `const.EXTRA_COLUMNS_DEST` is populated, If no, proceed to step 2, if yes, log and return.
2. iterate over each row and extract matches from `combined_os_column` using `const.EXTRA_COLUMNS_NON_WINDOWS_REGEX`
3. iterate over each row and extract matches from `combined_os_column` using `const.EXTRA_COLUMNS_WINDOWS_SERVER_REGEX`
4. iterate over each row and extract matches from `combined_os_column` using `const.EXTRA_COLUMNS_WINDOWS_DESKTOP_REGEX`
5.  iterate over each row and copy the value from the first result in that row from step 3 if the first result of step 2 is null.
6.  iterate over each row and copy the value from the first result in that row from step 4 if the first result of step 2 and step 3 is null.
7.  iterate over each row and copy the value from the second result in that row from step 3 if the second result of step 2 is null.
8.  iterate over each row and copy the value from the second result in that row from step 4 if the second result of step 2 and step 3 is null.
9.  iterate over each row and copy the value from the third result in that row from step 3 if the third result of step 2 is null.
10.  iterate over each row and copy the value from the third result in that row from step 4 if the third result of step 2 and step 3 is null.
11.  iterate over each row and copy the value from the original `combined_os_column` if the first result of step 2,3 and 4 is null.

Using the above explanation,  we iterate over every row 10 times.

Line numbers of that workflow in the original (red):

1: 89
2: 90-92
3: 93-95
4: 96-98
5: 101-103
6: 104-106
7: 101-103
8: 104-106
9: 101-103
10: 104-106
11: 107-110

New logic can be described as

1. check if `const.EXTRA_COLUMNS_DEST` is populated, If no, proceed to step 2, if yes, log and return.
2. iterate over each row and set matches from `combined_os_column` using `const.EXTRA_COLUMNS_NON_WINDOWS_REGEX` to `const.EXTRA_COLUMNS_DEST`
3. iterate over only rows where `const.EXTRA_COLUMNS_DEST` are still null and  set matches from `combined_os_column` using `const.EXTRA_COLUMNS_WINDOWS_SERVER_REGEX ` to `const.EXTRA_COLUMNS_DEST` 
4. iterate over only rows where `const.EXTRA_COLUMNS_DEST` are still null and  set matches from `combined_os_column` using `const.EXTRA_COLUMNS_WINDOWS_DESKTOP_REGEX ` to `const.EXTRA_COLUMNS_DEST` 
5.  iterate over only rows where `const.EXTRA_COLUMNS_DEST[0]` are still null and  and set to original `combined_os_column`

original step 2 is unchanged
original steps 3, 5, 7, and 19 are combined into the new step 3
original steps 4, 6, 8, and 10 are combined into the new step 4
original step 11 shifts to new step 5.

additionally,  all iterations after step 2 only iterate over rows where the relevant fields are null, since that is indexed.  Even if this wasn't the case,  the original required iterating over the entire dataset 10 times,  the new now iterates over the entire set 4 times.

In theory,  there is a minor logic difference in that if one group (os_name, version, arch) was extracted from one regex and another from another regex,  they could both be applied, but that both isn't the intention of the code and isn't actually possible given the regexs in question.  The new code basically treats the three groups as a tuple for the null check,  so all three must be null for the whole to be null.